File: player.py
from game import Game
from enum import Enum
import math
import util


class Weapon(Enum):
    knife = 1
    pistol = 2
    smg = 3
    chaingun = 4


class InputData:
    def __init__(self, x_dir, y_dir, weapon, is_strafe, is_fire, is_use):
        assert x_dir in {-1, 0, 1}
        assert y_dir in {-1, 0, 1}
        assert weapon is None or type(weapon) == Weapon
        assert type(is_strafe) == type(is_fire) == type(is_use) == bool
        self.x_dir = x_dir
        self.y_dir = y_dir
        self.weapon = weapon
        self.is_strafe = is_strafe
        self.is_fire = is_fire
        self.is_use = is_use

    def __setattr__(self, key, value):
        raise RuntimeError("InputData is immutable")


class State(Enum):
    simple = 0
    attack = 1


class Player:
    def __init__(self):
        self.state = State.simple
        self.face_count = 0
        self.weapon = self.best_weapon = Weapon.knife
        self.ammo = 0
        self.angle = 0
        self.c_rotate_speed = 5
        self.c_move_speed = 14999
        self.c_move_back_speed = 9999

    def do_actor(self, game: Game, input_data: InputData):
        if self.state == State.simple:
            self.act_simple(game, input_data)
        elif self.state == State.attack:
            self.act_attack(game, input_data)

    def act_simple(self, game: Game, input_data: InputData):
        self.update_face(game)
        self.change_weapon(game, input_data)
        if input_data.is_use:
            self.cmd_use(game)
        if input_data.is_fire:
            self.cmd_fire(game)
        self.move(game, input_data)

    def act_attack(self, game: Game, input_data: InputData):
        self.update_face(game)
        self.move(game, input_data)
        self.do_attack(game)

    def update_face(self, game: Game):
        self.face_count += 1
        if self.face_count > game.gen_rnd():
            self.face_count += 1
            game.gen_rnd()

    def change_weapon(self, game: Game, input_data: InputData):
        w = input_data.weapon
        if w is not None and self.ammo > 0 and self.best_weapon >= w:
            self.weapon = w

    def cmd_use(self, game: Game):
        # NOT IMPLEMENTED
        pass

    def cmd_fire(self, game: Game):
        # NOT IMPLEMENTED
        pass

    def move(self, game: Game, input_data: InputData):
        # x move
        if input_data.is_strafe:
            dx = input_data.x_dir * self.c_move_speed
            self.thrust(self.angle - 90, dx, game)
        else:
            self.angle += input_data.x_dir * self.c_rotate_speed

        # y move
        if input_data.y_dir == 1:
            dy = self.c_move_speed
        else:
            dy = -self.c_move_back_speed
        self.thrust(self.angle, dy, game)

    def thrust(self, angle: int, dist: int, game: Game):
        dx = int(math.floor(dist * util.cos_table[angle]))
        dy = -int(math.floor(dist * util.sin_table[angle]))
        self.clip_move(dx, dy, game)

    def clip_move(self, dx: int, dy: int, game: Game):
        if self.try_move(dx, dy, game):
            return None
        if self.try_move(dx, 0, game):
            return None
        if self.try_move(0, dy, game):
            return None

    def try_move(self, dx: int, dy: int, game: Game):
        # NOT IMPLEMENTED
        pass

    def do_attack(self, game: Game):
        # NOT IMPLEMENTED
        pass

    @staticmethod
    def correct_angle(angle, maxval=360):
        while angle < 0:
            angle += maxval
        while angle >= maxval:
            angle -= maxval

File: src/game/player.cpp
#include "player.h"
#include "game.h"
#include "consts.h"

#include <cassert>

void Player::do_actor(Game *game, const ControlInput &input) {
//    update_face(game);
//    if (state == PS_BASIC) {
//        change_weapon(game, input);
//        if (input.is_use) {
//            cmd_use(game);
//        }
//        else if (input.is_attack) {
//            cmd_fire(game);
//        }
//    }
    move(game, input);
//    if (state == PS_ATTACK) {
//        do_attack(game, input);
//    }
}

void Player::update_face(Game *game) {
}

void Player::change_weapon(Game *game, const ControlInput &input) {
    assert(input.weapon >= 0 and input.weapon <= Weapon::W_CHAINGUN);

    if (input.weapon != 0 and ammo > 0 and input.weapon <= best_weapon) {
        weapon = static_cast<Weapon>(input.weapon);
    }
}

void Player::move(Game *game, const ControlInput &input) {
    assert(-1 <= input.dx and input.dx <= 1);
    assert(-1 <= input.dy and input.dy <= 1);

    if (input.is_strafe) {
        int dx = input.dx * move_speed;
        thrust(game, angle - 90, dx);
    }
    else {
        angle += input.dx * rotate_speed;
        angle = normalize_angle(angle);
    }

    if (input.dy == 1) {
        thrust(game, angle, move_speed);
    }
    else if (input.dy == -1) {
        thrust(game, angle + 180, move_back_speed);
    }
}

int Player::normalize_angle(int angel) {
    while (angel < 0) {
        angel += 360;
    }
    while (angel >= 360) {
        angel -= 360;
    }
    return angel;
}

void Player::thrust(Game *game, int angle, int dist) {
    angle = normalize_angle(angle);
    int dx = floor(dist * COS_TABLE[angle]);
    int dy = floor(dist * SIN_TABLE[angle]);
    clip_move(game, dx, dy);
}

void Player::clip_move(Game *game, int dx, int dy) {
    try_move(game, dx, dy) or try_move(game, dx, 0) or try_move(game, 0, dy);
}

bool Player::try_move(Game *game, int dx, int dy) {
    int new_x = x + dx;
    int new_y = y + dy;

    int txl = (new_x - PLAYER_SIZE) >> TILE_SHIFT;
    int txh = (new_x + PLAYER_SIZE) >> TILE_SHIFT;
    int tyl = (new_y - PLAYER_SIZE) >> TILE_SHIFT;
    int tyh = (new_y + PLAYER_SIZE) >> TILE_SHIFT;

    for (int tx = txl; tx <= txh; tx++) {
        for (int ty = tyl; ty <= tyh; ty++) {
            if (game->map[tx][ty] == T_SOLID) {
                return false;
            }
        }
    }


    if (txl > 0) txl--;
    if (tyl > 0) tyl--;
    if (txh < game->map_size - 1) txh++;
    if (tyh < game->map_size - 1) tyh++;

    // Check if there are actor

    this->x = new_x;
    this->y = new_y;
    return true;
}
